### PR TITLE
feat: personal collection mgmt

### PR DESCRIPTION
**Pull request by poontology**
Hi, I was left thinking about comments in PR #8 ("would be nice if this project could also associated already downloaded dumps of zip/cbz files or folders so you could either read stuff directly off it just like LRR does") still haven't looked into LRR but came up with lightweight solution that's sufficient for my use but does require a little bit of setup from the user.

The screenshot shows the UI elements added, when gallery is marked as "Have" it gets flagged in gallery_personal db table and the UI shows a "Open cbz" link with a custom protocol and the gid "panda://12345", then the OS is configured to handle the custom protocol by running a script, the script knows the directories where user has the cbz files and can match the right file using the gid in the filename then it opens the reader app for that file.

Screenshot

To setup the protocol handler in linux create the script and the handler config:

    cat <<EOF >~/.local/bin/panda.sh
    #!/bin/bash
    [[ ! "$1" =~ ^panda://[0-9]+$ ]] && echo "Syntax: $0 panda://ID" && exit 1
    ID="${1#panda://}"
    # ID followed by one non-numeric character to prevent partial match
    FN=$(find /media/hentai/ -name "${ID}[^0-9]*cbz" -print -quit)
    [ -f "$FN" ] && evince "$FN"
    EOF
    chmod a+x ~/.local/bin/panda.sh

    cat <<EOF >~/.local/share/applications/panda-handler.desktop
    [Desktop Entry]
    Name=panda
    Exec=panda.sh %u
    Type=Application
    MimeType=x-scheme-handler/panda;
    EOF

    # register handler
    xdg-mime default panda-handler.desktop x-scheme-handler/panda
    # this should print the handler filename
    xdg-mime query default x-scheme-handler/panda
    # update browsers etc
    update-desktop-database ~/.local/share/applications/
    # cli open test
    xdg-open panda://123123

That's enough for it to open cbz files from pandabrowser but chrome keeps asking to verify each click which is annoying so by opening chrome://policy/ and "More actions" -> "View logs" shows what path it's loading managed policies from, for me chromium uses /etc/chromium-browser/policies/managed/ and if you create a managed_policies.json file that says `{
 "URLAllowlist": ["panda://*"]
}` in that path it allows the browser to open those links without asking anything.

Once that's all done user could have another script that identifies their cbz files with /imageSearch, renames them with the gid (maybe also adds /comicInfo file in the cbz) and sends a /personal API request to flag the gallery as something they have, then they can search "have" galleries and click to open.

All that's of course entirely different on windows and macos so if this feature is included maybe those should be looked up as well. Also wasn't too sure if everybody necessarily wants this feature so added a feature flag in config.js that hides the UI changes.

Thoughts?

### CODE REVIEW

**Comment by UBTL** [comment]
Alright I was finally make some time to personally test everything up till now AND..... everything from my side of thing is utterly broken even with the applied databases changes and the like..... so I would like to request a copy of your database that should work with all these changes, not a starter database but either the old 2024_01 database or the newer 2024_07 if you happened to be working on that version that I shared since I cant seem to use my current database versions with the updated repo.

Talking about the 2024_07 DB related to the topic at hand here, first of all the dumps of zip/cbz files I mentioned is related to my UBUCA releases https://sukebei.nyaa.si/view/4058234 which I originally used LRR to "view" but Im sick how fucking broken LRR can most of the times and I wanted an alternative to it so I was thinking in maybe PandaBrowser could work similarish like a gallery viewer too from already downloaded files, basically either it could scan a folder or the user could manually add the path to the file/folder where the files are when you click on the gallery button or whatever, it would open the page/gallery of said entry, also suggested way here regarding how to handle already downloaded cbz files is that those need to start with the gid number first and most people me included dont make our files starting with such name scheme usually however try to match the file names with the very same is highly unlikely and quite prone to many wrongly associated galleries and the works so one option was have the user manually add the path to the said gallery/file themselves another possible option making use of the new thumbnail hashing perhaps make use of that to associate possible galleries automatically? there may be some galleries still wrongly associated but when we are dealing with massive amounts of entries like here that cover 16 years worth of content, its inevitable I guess and beats have to manually associate files/galleries specially when you hav tens of thousand of files pre-downloaded already.

prototype 2

the inclusion of ComicInfo.xml is great though, I happened to talk with the admin of [HenTag ](https://hentag.com/) about providing a dump of all the metadata files the community has developed over there and he said sure he was going to..... that was 2 fucking years ago and some months he bitched about having some issues regarding the servers of this project and he hasnt yet feel fucked to release any of them when I mentioned again

**Comment by poontology** [comment]
tried out the 09/27 dump, after `docker-compose up` on first run of `npx dbmate --url 'mysql://root:69@localhost/panda' up` it said "Error 1062 (23000): Duplicate entry '0' for key 'PRIMARY'" because I'd somehow managed to leave out the auto_increment flag on the id field and forgot to test the thing from scratch, fixed now in this branch.

also in the new db dump gallery-table is set to InnoDB, maybe it's ok with gigabytes of memory but with my default settings under docker it's miserably slow to write to (seems like it wants to spend hours just collecting row-locks before even thinking about writing something), so switched it back to MyISAM.

after those two changes this is what rest of bringing up the new dump looks like for me:

    $ npx dbmate --url 'mysql://root:69@mariadb/panda' status
    [X] 20240101000627_initial-schema.sql
    [X] 20240925004201_thumbnail-schema.sql
    [ ] 20240925005204_thumbnail-data.sql
    [ ] 20240929110332_gallery-personal-schema.sql

    Applied: 2
    Pending: 2

    # rollback the table with bad id-field
    $ npx dbmate --url 'mysql://root:69@mariadb/panda' down
    Rolling back: 20240925004201_thumbnail-schema.sql
    Rolled back: 20240925004201_thumbnail-schema.sql in 14.54028867s

    # and apply again
    $ npx dbmate --url 'mysql://root:69@mariadb/panda' up
    Applying: 20240925004201_thumbnail-schema.sql
    Applied: 20240925004201_thumbnail-schema.sql in 15.780869575s
    Applying: 20240925005204_thumbnail-data.sql
    Applied: 20240925005204_thumbnail-data.sql in 10m49.841633636s
    Applying: 20240929110332_gallery-personal-schema.sql
    Applied: 20240929110332_gallery-personal-schema.sql in 3.745791ms

(forgot to say in previous PR that it reads the db credentials from `.env` file but I guess you figured it out, perhaps it'd still be more obvious to be documented with the --url argument in README)

gallery-table seems to be missing keys and is bigger so it took 10min instead of 3min on the january dump, also tag search doesn't work without `posted` index so I did `alter table gallery add primary key (gid); alter table gallery add key (posted);`. with those things seem to mostly work with the new dump although many of the new galleries have their tags in text fields in the gallery table, should they be inserted to gid_tid? I can upload the db to gdrive if you want but should we do some clean up?

also applied the new thumbnails on top of the old ones with some python (for reference):

    from glob import glob
    from os.path import basename
    oldlst=set([basename(f) for f in glob('pandathumbs/**/*jpg')])
    newlst=[basename(f) for f in glob('2024/**/*jpg')]
    len(oldlst.intersection(newlst))
    ## 5922 imgs were already part of the january dump
    ## checked that the dupes matched in size so just overwrote them..
    import os
    TDIR='/blabla/PandaBrowser/frontend/pandathumbs'
    for f in newlst: os.rename(f, f"{TDIR}/{f[0:2]}/{f[2:4]}/{f}")

from a bit of browsing I guess many of the 85k thumbs since 2024-05-05 are still to be downloaded?

**Comment by poontology** [comment]
@UBTL (thought the PR was still a little half-cooked so had it marked as draft to maybe talk about it but everything can of course be changed later)

re: reader functionality, there will no doubt have to be compromises no matter the approach. personally I probably wouldn't use a reader through a browser but yea one could no doubt be added, would just need to figure out the scanning and thumbnailing process.

one reason why I liked the id in filename approach when used to launch a native reader specifically is the adaptability/independence, user can edit the launch script to point it to their own media folders, match the id from filename using their own naming schemes (eg. "filename [eh-123].cbz" doesn't need to be just number prefix) without making their file collection dependent on a system that they might not be 100% sold on long term.
but, in case the system wants to be aware of all the album files then a standalone launcher script wouldn't be enough.

just spitballing but I guess a scanner script could collect the cbz file paths, page/image count and maybe hash first three images from the album (because the gallery cover may not always be the first image in the album, of course ehentai has hashed every image of every gallery for their file search feature but I think that's probably unnecessary for a single-user system). with those collected in the db perhaps there wouldn't need to be a direct file<->gallery mapping, the gallery UI could just suggest that given gallery seems to match a file if the user wants to look at it (multiple galleries might match the same file)?

**Comment by UBTL** [comment]
I was wondering why stuff was broken, found out the SQL files had conflicting instructions (for the way I load them up) fucking up royally the database at times until I went and manually read each file, made a new database that should have the new schema working and up over https://drive.google.com/drive/folders/1gkOQD2jMDmubbzmHdZk32OKkNV_rdzXU?usp=sharing as well as the downloaded thumbnails with sorted on their folders and such, I still dont know why after the January dump pretty much every most of the entries has broken thumbnails, I added the the dumped urls perhaps to find out all the supposed missing thumbnails over here.

New version of PandaBrowser is really quite faster, I actually redownloaded all of this stuff back on from my archives since I deleted all of this on my main PC since Im severely space starved lately and running what I would name version 1(?) of PB,  damn it was slow as dog shit that is bordering on unusable, tried the image search feature with thumbnails of stuff I downloaded but doesnt seem to yet work on my side so far.

Regarding the Tags, I have no idea of all the new tags added from Jan 1st up till now so the tag data is in there for now, kind of the same regarding the torrent data too it is embedded on the gallery table rather then the torrent table.

After I sleep some I gave it a better thought and yeah I may be not push for the idea of the viewer anymore.... also I found out https://github.com/ComicLib/comiclib which seems to be like LRR but that doesnt such massively so yeah, dont bother much with that imho, though I would do like to have the option to add some custom tags on the galleries specially for the tags that Tenboro has refused forever to approve, specially the most voted "Vanilla" tag, could be added to the "Others: " tag category those ones

The Open CBZ feature can be still done as a feature for those willing to change their CBZ files and add a gid into their filename and have the companion script or instruction on how to associate some comic reader to the command and the like if anyone wishes for it.

**Comment by UBTL** [comment]
One option to compare what new tags there are would be check up on https://ehwiki.org/wiki/Category:Tag
other would be comb through https://github.com/EhTagTranslation/Database

**Comment by UBTL** [comment]
Actually I just finished checking properly and is from page 8031 and lower where all the thumbnails break inevitably
E-Hentai-DB

**Comment by UBTL** [comment]
Alright I uploaded to the drive folder what should be most of the missing thumbnails for now I guess, Im sure there most be still some missing but most should be there now

**Comment by poontology** [comment]
> found out the SQL files had conflicting instructions (for the way I load them up)

you ran them manually? yea they aren't supposed to be used like that. each file contains the "up" part which is how the change is applied and "down" part which says how it's rolled back (removed) so they need to be applied with the program that manages the migrations, dbmate (see previous comment for the commands). 
this is how changes to databases are managed in software projects because the system could be running in lots of environments which evolve at different pace managed by different people so to bring up the database to a known state that matches what the code is expecting of it needs to be done in a structured way.

I mentioned in the db change PR that it keeps it's state in the schema_migrations table so if some of the changes have been done manually you might have to either undo those to successfully run the `dbmate up` command but it might be easier to restore from a dump though.

edit: on second thought maybe easiest if I just upload a new db dump on gdrive, give me a day or so I'll clean up the gallery table while at it, move the tags to the tag tables etc.

**Comment by UBTL** [comment]
Hello again, If dbmate was mentioned I completely glossed over that part when going through the messages here and didnt catch that is how it was being used, I personally use either mysql-cli or HeidiSQL to deal with mysql stuff so I didnt know about that stuff that is why when loading the files from either of those it really caused me issues, sorry then, need to make the annotation to the readme about it for others that may be like me on that regard when they deal with mysql files.
Regarding the feature of the custom protocol stuff perhaps to not deal with custom scripts and the like SO wise, maybe use something like this? https://www.npmjs.com/package/protocol-registry , can still add the custom script to those wishing to use that more manual approach or for them to customize it to their desires also perhaps mention this module too if it can work for this usages https://www.npmjs.com/package/react-book-reader 

Also yeah if you could drop me a copy of your database or how it should be made I guess please.

**Comment by poontology** [comment]
yeah could maybe put a separate README file in the db/migrations/ dir just to make it clearer what they're for. (for additional context: think of dbmate a bit like git, git can upload and download files but that's not what it's for exactly, it's for version control management, dbmate can execute arbitrary sql but what it's for specifically is to keep track of what changes have been applied to db and then import only new changes from new code releases.. for anything else, like restoring dump files you'd still use regular old sql client). if regular user downloads a db dump with a compatible app version then they don't need to know anything about these things since they have up-to-date db, applying migrations is just for people who want to do incremental code updates.

nice looks like protocol-registry would simplify having to document that part of setup. at least for linux I see it does what I had in the initial comment, creates a .desktop file and runs xdg-mime (making browser open the links without prompts would still require a manual policy change though). the open link as it is now though only has the gallery `gid` so there would need to be something like the example launcher script that would know where to find a file matching the gid and what to open it with. to be able to open the file without a script the system would have to know where all the users' files are, to match them to galleries, to keep scanning them in case they move the files around, and they would need to be re-added if user updates the db from new dump.

**Comment by poontology** [comment]
@UBTL so, went over the text columns in the new db for anything that was added after the january dump (gid>2783144 /2023-12-31), inserted 4659 new tags that weren't yet in tag-table, linked the new galleries to all the tags from the text columns and inserted the 267674 torrents those galleries had to the torrent-table.
also because the tag-table unique key was also covered id it wasn't protecting uniqueness and there were duplicate tags in the table, cleaned those out and replaced the unique key.

there's still some things to look at.. there are 5229 tags not used for anything `select count(*) from tag where id not in (select distinct tid from gid_tid);` most (5148) of which are non-typed tags `..and name not like '%:%'`.

checked the thumb zips, from pandathumbs-2024-missing.zip 60k new imgs were added (8511 of them were already in) and all of pandathumbs-July-2024.zip were already in (probably in the 2024-07-02 zip).
now we have 119504 imgs that aren't linked to any galleries `select (select count(*) from thumbnail) - (select count(distinct thumbnail_id) from gallery) unusedimgs;` and 27853 galleries that point to a local file that's missing (at least from my drive) `select count(*) from thumbnail where ph0 is null and thumb not like '%/%';`

the db dump is at https://drive.google.com/drive/folders/1aRiXKdoPF-32H3Y7iNsdsi6qWWKhnwzV?usp=drive_link exported pandadb-20241010-1.sql.gz last night but then noticed that if I click a gid in UI it wouldn't actually find the gallery with it's own gid, looked at the query and it's actually searching for root_gid and they were something different on the september dump than they were on january dump. so copied root_gids from january db to this db and while I'm not sure what is in any of the newer galleries for that field at least the old gid searches work now, made a new pandadb-20241010-2.sql.gz export from that.

where was the data in the text columns btw from? it looked like python dicts (not valid json because of single-quotes). is there a process to fetch the new data?

**Comment by UBTL** [comment]
@poontology great job amazing, thank you for dealing with all of these
If you can get the 27853 missing thumbnails then I guess I could dump them once again, I though they would be the Private Galleries thumbnails http://localhost:6969/?category=1024  but those would only be 4K thumbnails, still way far off from the final amount talking about root_gid the torents arent showing up on the entries because it is missing that but that is easy fix with " UPDATE gallery SET root_gid=gid; " and the other fields used
added = unix time
addedstr = human readable date string
fsize = filesize in bytes
fsizestr = filesize in readable form

> where was the data in the text columns btw from? it looked like python dicts (not valid json because of single-quotes). is there a process to fetch the new data?

The original dump was done by Sachia Lanlus back in July of 2019 when panda went down for the first time https://mega.nz/folder/oh1U0SIA#WBUcf3PaOvrfIF238fnbTg
After that [gipaf23445](https://sukebei.nyaa.si/user/gipaf23445) has been dropping database dumps of the site from time to time they are in sqlite form though, from his database dumps is where I have hacked (in the literally hack apart and paste together) this barely put together MYSQL database of my first release back in January, as I mentioned before I forgot completely how I build that database and I didnt documented nothing about it so when the never dump was uploaded back in July , I just converted the database into MYSQL and tried to piece it together into something workable for PandaBrowser as well as dump the new batch of thumbnails and such off it, Im sure gipaf23445 must have some sort of python script that is constantly pinging the API to record new entries from time to time thus why it is formatted that way, after all he is the one that is keeping documentation of all the works DMCAed off Panda in here https://calc.idash.org/ehdmca

> TL;DR : The original database dumps arent mine, I just converted it into MYSQL

**Comment by poontology** [comment]
@UBTL oh right saw those in sukebei and even downloaded the sqlite then forgot to take a closer look when I started playing around with this project. I think they're fetching EH updates with a python script (there's a .pickle file in the mega drive which is a python object serialization format) just for some reason they've saved str(dict) in the db instead of json which would've been more useful also because sqlite supports it.

I wrote couple scripts to first fix the data into json format and then a cleaner script that reorganizes the db into table structure similar to pandabrowser's db (the tag linking sql is quite a pile of cross joins but runs surprisingly quick). scripts and the sqlite db are in the gdrive. the sqlite has almost 3x the number of torrents and 30% more linked tags to galleries perhaps it could be used as a new base.
